File: clients/typescript/src/http/client.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest'
import { createHTTPClient, HTTPError } from './client.js'
import type { HTTPConfig } from './client.js'
import type { Flag } from '../types.js'

// Helpers

const baseConfig: HTTPConfig = {
  baseURL: 'http://localhost:8080',
  apiKey: 'test-id.test-secret',
}

const sampleFlag: Flag = {
  key: 'my-flag',
  description: 'A test flag',
  enabled: true,
}

function makeResponse(body: unknown, status = 200): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { 'Content-Type': 'application/json' },
  })
}

function mockFetch(response: Response | (() => Response)) {
  const fn = vi.fn().mockImplementation(() =>
    typeof response === 'function' ? response() : response,
  )
  return fn
}

function captureRequests(): [typeof globalThis.fetch, { url: string; init?: RequestInit }[]] {
  const calls: { url: string; init?: RequestInit }[] = []
  const fn = vi.fn().mockImplementation((url: string, init?: RequestInit) => {
    calls.push({ url, init })
    return Promise.resolve(makeResponse({ flag: { key: 'x', enabled: true } }))
  }) as typeof globalThis.fetch
  return [fn, calls]
}

// -- FlagManager tests -------------------------------------------------------

describe('createHTTPClient', () => {
  describe('createFlag', () => {
    it('POSTs to /v1/flags and returns mapped flag', async () => {
      const fetch = mockFetch(makeResponse({ flag: { key: 'my-flag', enabled: true, created_at: '2024-01-01T00:00:00Z' } }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      const result = await client.createFlag(sampleFlag)
      expect(result.key).toBe('my-flag')
      expect(result.enabled).toBe(true)
      expect(result.createdAt).toBe('2024-01-01T00:00:00Z')
      expect(fetch).toHaveBeenCalledWith(
        'http://localhost:8080/v1/flags',
        expect.objectContaining({ method: 'POST' }),
      )
    })

    it('sends Authorization header', async () => {
      const [fetch, calls] = captureRequests()
      const client = createHTTPClient({ ...baseConfig, fetch })
      await client.createFlag(sampleFlag)
      expect(calls[0].init?.headers).toMatchObject({ Authorization: 'Bearer test-id.test-secret' })
    })
  })

  describe('getFlag', () => {
    it('GETs /v1/flags/:key', async () => {
      const fetch = mockFetch(makeResponse({ flag: { key: 'my-flag', enabled: false } }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      const result = await client.getFlag('my-flag')
      expect(result.key).toBe('my-flag')
      expect(fetch).toHaveBeenCalledWith(
        'http://localhost:8080/v1/flags/my-flag',
        expect.objectContaining({ method: 'GET' }),
      )
    })

    it('throws HTTPError on 404', async () => {
      const fetch = vi.fn().mockResolvedValue(new Response('not found', { status: 404 }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      await expect(client.getFlag('missing')).rejects.toThrow(HTTPError)
      await expect(client.getFlag('missing')).rejects.toMatchObject({ status: 404 })
    })

    it('throws HTTPError on 401', async () => {
      const fetch = vi.fn().mockResolvedValue(new Response('unauthorized', { status: 401 }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      await expect(client.getFlag('x')).rejects.toMatchObject({ status: 401 })
    })
  })

  describe('listFlags', () => {
    it('GETs /v1/flags and returns all flags', async () => {
      const fetch = mockFetch(
        makeResponse({ flags: [{ key: 'a', enabled: true }, { key: 'b', enabled: false }] }),
      )
      const client = createHTTPClient({ ...baseConfig, fetch })
      const flags = await client.listFlags()
      expect(flags).toHaveLength(2)
      expect(flags[0].key).toBe('a')
    })

    it('returns empty array when flags is null/absent', async () => {
      const fetch = mockFetch(makeResponse({ flags: null }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      const flags = await client.listFlags()
      expect(flags).toEqual([])
    })
  })

  describe('updateFlag', () => {
    it('PUTs to /v1/flags/:key', async () => {
      const fetch = mockFetch(makeResponse({ flag: { key: 'my-flag', enabled: false } }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      const result = await client.updateFlag({ ...sampleFlag, enabled: false })
      expect(result.enabled).toBe(false)
      expect(fetch).toHaveBeenCalledWith(
        'http://localhost:8080/v1/flags/my-flag',
        expect.objectContaining({ method: 'PUT' }),
      )
    })
  })

  describe('deleteFlag', () => {
    it('DELETEs /v1/flags/:key', async () => {
      const fetch = vi.fn().mockResolvedValue(new Response(null, { status: 204 }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      await expect(client.deleteFlag('my-flag')).resolves.toBeUndefined()
      expect(fetch).toHaveBeenCalledWith(
        'http://localhost:8080/v1/flags/my-flag',
        expect.objectContaining({ method: 'DELETE' }),
      )
    })
  })

  // -- Evaluator tests -------------------------------------------------------

  describe('evaluate', () => {
    it('POSTs to /v1/evaluate with single key', async () => {
      const fetch = vi.fn().mockResolvedValue(makeResponse({ results: [{ key: 'my-flag', value: true }] }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      const v = await client.evaluate('my-flag', {}, false)
      expect(v).toBe(true)
      const body = JSON.parse(fetch.mock.calls[0][1]?.body as string)
      expect(body.key).toBe('my-flag')
    })

    it('sends Authorization header', async () => {
      const fetch = vi.fn().mockResolvedValue(makeResponse({ results: [{ key: 'x', value: false }] }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      await client.evaluate('x', {}, false)
      expect(fetch.mock.calls[0][1]?.headers).toMatchObject({ Authorization: 'Bearer test-id.test-secret' })
    })
  })

  describe('evaluateBatch', () => {
    it('POSTs requests array to /v1/evaluate', async () => {
      const fetch = vi.fn().mockResolvedValue(
        makeResponse({ results: [{ key: 'a', value: true }, { key: 'b', value: false }] }),
      )
      const client = createHTTPClient({ ...baseConfig, fetch })
      const results = await client.evaluateBatch([
        { key: 'a', defaultValue: false },
        { key: 'b', defaultValue: true },
      ])
      expect(results).toEqual([{ key: 'a', value: true }, { key: 'b', value: false }])
      const body = JSON.parse(fetch.mock.calls[0][1]?.body as string)
      expect(Array.isArray(body.requests)).toBe(true)
      expect(body.requests).toHaveLength(2)
    })
  })

  // -- SSE streaming tests ---------------------------------------------------

  describe('stream', () => {
    function makeSSEResponse(events: string): Response {
      const encoder = new TextEncoder()
      const encoded = encoder.encode(events)
      const stream = new ReadableStream({
        start(controller) {
          controller.enqueue(encoded)
          controller.close()
        },
      })
      return new Response(stream, {
        status: 200,
        headers: { 'Content-Type': 'text/event-stream' },
      })
    }

    it('yields update and delete events from SSE stream', async () => {
      const sseData = [
        'id:1\nevent:update\ndata:{"key":"flag-a","enabled":true}\n\n',
        'id:2\nevent:delete\ndata:{"key":"flag-b"}\n\n',
      ].join('')

      const fetch = vi.fn().mockResolvedValue(makeSSEResponse(sseData))
      const client = createHTTPClient({ ...baseConfig, fetch })

      const events = []
      for await (const ev of client.stream()) {
        events.push(ev)
      }

      expect(events).toHaveLength(2)
      expect(events[0]).toMatchObject({ type: 'update', key: 'flag-a', eventId: '1' })
      expect(events[1]).toMatchObject({ type: 'delete', key: 'flag-b', eventId: '2' })
    })

    it('sends Last-Event-ID header on reconnect', async () => {
      const fetch = vi.fn().mockResolvedValue(makeSSEResponse(''))
      const client = createHTTPClient({ ...baseConfig, fetch })
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      for await (const _ of client.stream('42')) { break }
      expect(fetch.mock.calls[0][1]?.headers).toMatchObject({ 'Last-Event-ID': '42' })
    })

    it('throws HTTPError on non-2xx SSE response', async () => {
      const fetch = vi.fn().mockResolvedValue(new Response('forbidden', { status: 403 }))
      const client = createHTTPClient({ ...baseConfig, fetch })
      await expect(async () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        for await (const _ of client.stream()) { /* no-op */ }
      }).rejects.toThrow(HTTPError)
    })

    it('sends Authorization header on stream request', async () => {
      const fetch = vi.fn().mockResolvedValue(makeSSEResponse(''))
      const client = createHTTPClient({ ...baseConfig, fetch })
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      for await (const _ of client.stream()) { /* no-op */ }
      expect(fetch.mock.calls[0][1]?.headers).toMatchObject({ Authorization: 'Bearer test-id.test-secret' })
    })

    it('concatenates multi-line data fields into a single JSON object', async () => {
      // SSE spec: consecutive data: lines are concatenated with '\n'.
      // Split a valid JSON object across two data: lines to verify concatenation.
      const sseData = 'event:update\ndata:{"key":"multi-flag",\ndata:"enabled":true}\n\n'
      const fetch = vi.fn().mockResolvedValue(makeSSEResponse(sseData))
      const client = createHTTPClient({ ...baseConfig, fetch })
      const events = []
      for await (const ev of client.stream()) {
        events.push(ev)
      }
      expect(events).toHaveLength(1)
      expect(events[0]).toMatchObject({ type: 'update', key: 'multi-flag' })
    })
  })
})

// Ensure the module doesn't import @grpc/* types at the top level.
// (If it did, this test file would fail to load without @grpc/grpc-js installed.)
it('module imports without gRPC dependency', async () => {
  const mod = await import('./client.js')
  expect(typeof mod.createHTTPClient).toBe('function')
})

File: clients/go/http/client_test.go
package http_test

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	flagz "github.com/matt-riley/flagz/clients/go"
	flagzhttp "github.com/matt-riley/flagz/clients/go/http"
)

// helpers

func mustMarshal(v any) []byte {
	b, _ := json.Marshal(v)
	return b
}

func flagJSON(key string, enabled bool) string {
	return fmt.Sprintf(`{"flag":{"key":%q,"description":"desc","enabled":%v,"variants":null,"rules":null,"created_at":"2024-01-01T00:00:00Z","updated_at":"2024-01-01T00:00:00Z"}}`, key, enabled)
}

func newTestServer(t *testing.T, handler http.HandlerFunc) (*httptest.Server, *flagzhttp.Client) {
	t.Helper()
	srv := httptest.NewServer(handler)
	t.Cleanup(srv.Close)
	c := flagzhttp.NewHTTPClient(flagzhttp.Config{
		BaseURL: srv.URL,
		APIKey:  "test-key",
	})
	return srv, c
}

func assertAuth(t *testing.T, r *http.Request) {
	t.Helper()
	got := r.Header.Get("Authorization")
	if got != "Bearer test-key" {
		t.Errorf("auth header: got %q, want %q", got, "Bearer test-key")
	}
}

// -- CRUD tests --------------------------------------------------------------

func TestCreateFlag(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		assertAuth(t, r)
		if r.Method != http.MethodPost || r.URL.Path != "/v1/flags" {
			t.Errorf("unexpected %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, flagJSON("my-flag", true))
	})
	f, err := c.CreateFlag(context.Background(), flagz.Flag{Key: "my-flag", Enabled: true})
	if err != nil {
		t.Fatal(err)
	}
	if f.Key != "my-flag" || !f.Enabled {
		t.Errorf("unexpected flag: %+v", f)
	}
	if f.CreatedAt.IsZero() {
		t.Error("expected CreatedAt to be set")
	}
}

func TestGetFlag(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		assertAuth(t, r)
		if r.Method != http.MethodGet || r.URL.Path != "/v1/flags/my-flag" {
			t.Errorf("unexpected %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, flagJSON("my-flag", true))
	})
	f, err := c.GetFlag(context.Background(), "my-flag")
	if err != nil {
		t.Fatal(err)
	}
	if f.Key != "my-flag" {
		t.Errorf("got key %q", f.Key)
	}
}

func TestGetFlagNotFound(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "not found", http.StatusNotFound)
	})
	_, err := c.GetFlag(context.Background(), "missing")
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	var apiErr *flagzhttp.APIError
	if !isAPIError(err, &apiErr) || apiErr.StatusCode != http.StatusNotFound {
		t.Errorf("expected 404 APIError, got %v", err)
	}
}

func TestGetFlagUnauthorized(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
	})
	_, err := c.GetFlag(context.Background(), "x")
	var apiErr *flagzhttp.APIError
	if !isAPIError(err, &apiErr) || apiErr.StatusCode != http.StatusUnauthorized {
		t.Errorf("expected 401 APIError, got %v", err)
	}
}

func TestListFlags(t *testing.T) {
	// Use a simpler server for list
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, `{"flags":[{"key":"a","enabled":true},{"key":"b","enabled":false}]}`)
	}))
	defer srv.Close()
	cl := flagzhttp.NewHTTPClient(flagzhttp.Config{BaseURL: srv.URL, APIKey: "k"})
	flags, err := cl.ListFlags(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if len(flags) != 2 {
		t.Fatalf("want 2 flags, got %d", len(flags))
	}
}

func TestUpdateFlag(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		assertAuth(t, r)
		if r.Method != http.MethodPut || r.URL.Path != "/v1/flags/my-flag" {
			t.Errorf("unexpected %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, flagJSON("my-flag", false))
	})
	f, err := c.UpdateFlag(context.Background(), flagz.Flag{Key: "my-flag", Enabled: false})
	if err != nil {
		t.Fatal(err)
	}
	if f.Enabled {
		t.Error("expected Enabled=false")
	}
}

func TestDeleteFlag(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		assertAuth(t, r)
		if r.Method != http.MethodDelete || r.URL.Path != "/v1/flags/my-flag" {
			t.Errorf("unexpected %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(http.StatusNoContent)
	})
	if err := c.DeleteFlag(context.Background(), "my-flag"); err != nil {
		t.Fatal(err)
	}
}

// -- Evaluate tests ----------------------------------------------------------

func TestEvaluate(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		assertAuth(t, r)
		if r.Method != http.MethodPost || r.URL.Path != "/v1/evaluate" {
			t.Errorf("unexpected %s %s", r.Method, r.URL.Path)
		}
		var body map[string]any
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Error(err)
		}
		if body["key"] != "my-flag" {
			t.Errorf("unexpected key: %v", body["key"])
		}
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, `{"key":"my-flag","value":true}`)
	})
	v, err := c.Evaluate(context.Background(), "my-flag", flagz.EvaluationContext{}, false)
	if err != nil {
		t.Fatal(err)
	}
	if !v {
		t.Error("expected true")
	}
}

func TestEvaluateBatch(t *testing.T) {
	_, c := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		assertAuth(t, r)
		var body map[string]any
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Error(err)
		}
		reqs, ok := body["requests"].([]any)
		if !ok || len(reqs) != 2 {
			t.Errorf("expected 2 requests, got %v", body["requests"])
		}
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, `{"results":[{"key":"a","value":true},{"key":"b","value":false}]}`)
	})
	results, err := c.EvaluateBatch(context.Background(), []flagz.EvaluateRequest{
		{Key: "a", DefaultValue: false},
		{Key: "b", DefaultValue: true},
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(results) != 2 || results[0].Key != "a" || !results[0].Value {
		t.Errorf("unexpected results: %+v", results)
	}
}

// -- SSE streaming tests -----------------------------------------------------

func TestStream(t *testing.T) {
	events := []string{
		"id:1\nevent:update\ndata:{\"key\":\"flag-a\",\"enabled\":true}\n\n",
		"id:2\nevent:delete\ndata:{\"key\":\"flag-b\"}\n\n",
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Authorization") != "Bearer test-key" {
			http.Error(w, "unauth", http.StatusUnauthorized)
			return
		}
		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")
		flusher := w.(http.Flusher)
		for _, ev := range events {
			fmt.Fprint(w, ev)
			flusher.Flush()
		}
	}))
	defer srv.Close()

	c := flagzhttp.NewHTTPClient(flagzhttp.Config{BaseURL: srv.URL, APIKey: "test-key"})
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	ch, err := c.Stream(ctx, 0)
	if err != nil {
		t.Fatal(err)
	}

	var received []flagz.FlagEvent
	for ev := range ch {
		received = append(received, ev)
	}

	if len(received) != 2 {
		t.Fatalf("want 2 events, got %d: %+v", len(received), received)
	}
	if received[0].Type != "update" || received[0].EventID != 1 {
		t.Errorf("event 0: %+v", received[0])
	}
	if received[1].Type != "delete" || received[1].EventID != 2 {
		t.Errorf("event 1: %+v", received[1])
	}
}

func TestStreamLastEventIDHeader(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		got := r.Header.Get("Last-Event-ID")
		if got != "42" {
			t.Errorf("Last-Event-ID: got %q, want %q", got, "42")
		}
		w.Header().Set("Content-Type", "text/event-stream")
		// No events; just close.
	}))
	defer srv.Close()

	c := flagzhttp.NewHTTPClient(flagzhttp.Config{BaseURL: srv.URL, APIKey: "k"})
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	ch, err := c.Stream(ctx, 42)
	if err != nil {
		t.Fatal(err)
	}
	for range ch {
	}
}

func TestStreamContextCancellation(t *testing.T) {
	done := make(chan struct{})
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/event-stream")
		flusher := w.(http.Flusher)
		flusher.Flush()
		// Hold open until the request context is cancelled.
		<-r.Context().Done()
		close(done)
	}))
	defer srv.Close()

	c := flagzhttp.NewHTTPClient(flagzhttp.Config{BaseURL: srv.URL, APIKey: "k"})
	ctx, cancel := context.WithCancel(context.Background())

	ch, err := c.Stream(ctx, 0)
	if err != nil {
		t.Fatal(err)
	}

	// Cancel after a brief moment.
	time.AfterFunc(100*time.Millisecond, cancel)

	// Channel should close without hanging.
	timeout := time.After(3 * time.Second)
	for {
		select {
		case _, ok := <-ch:
			if !ok {
				return // channel closed as expected
			}
		case <-timeout:
			t.Fatal("timed out waiting for stream channel to close")
		}
	}
}

// -- helpers -----------------------------------------------------------------

func isAPIError(err error, target **flagzhttp.APIError) bool {
	if err == nil {
		return false
	}
	if e, ok := err.(*flagzhttp.APIError); ok {
		*target = e
		return true
	}
	return false
}

// Ensure Client satisfies interfaces at compile time.
var _ flagz.FlagManager = (*flagzhttp.Client)(nil)
var _ flagz.Evaluator = (*flagzhttp.Client)(nil)
var _ flagz.Streamer = (*flagzhttp.Client)(nil)

// Ensure types are usable.
var _ = strings.TrimSpace
